File: ContactList/app/src/main/java/com/bergamin/contactlist/SendContactsTask.java
package com.bergamin.contactlist;

import android.app.ProgressDialog;
import android.content.Context;
import android.os.AsyncTask;
import android.widget.Toast;

import com.bergamin.contactlist.converter.ContactConverter;
import com.bergamin.contactlist.dao.ContactDAO;
import com.bergamin.contactlist.model.Contact;

import java.util.List;

/**
 * Created by Guilherme on 30/10/2016.
 */

/*
Generics use:
 - params in doInBackground()
 - used during execution
 - response in onPostExecute()
*/
public class SendContactsTask extends AsyncTask<Void, Void, String> {

    private Context context;
    private ProgressDialog dialog;

    public SendContactsTask(Context context){
        this.context = context;
    }

    @Override
    protected String doInBackground(Void... params) {
        ContactDAO dao = new ContactDAO(context);
        List<Contact> contacts = dao.getContacts();
        dao.close();
        ContactConverter converter = new ContactConverter();
        String json = converter.convertoToJSON(contacts);
        WebClient client = new WebClient();
        return client.post(json);
    }

    @Override
    protected void onPreExecute() {
        dialog = ProgressDialog.show(context,"Please wait","Sending Contacts...",true,true);
    }

    @Override
    protected void onPostExecute(String response) {
        dialog.dismiss();
        Toast.makeText(context,response,Toast.LENGTH_LONG).show();
    }
}

File: Kotlin/ContactList/app/src/main/java/com/bergamin/contactlist/util/SendContactsTask.kt
package com.bergamin.contactlist.util

import android.app.ProgressDialog
import android.content.Context
import android.os.AsyncTask
import android.widget.Toast
import com.bergamin.contactlist.R
import com.bergamin.contactlist.dao.ContactDAO
import com.google.gson.Gson

/**
 * Created by Guilherme Taffarel Bergamin on 29/11/2017.
 *
 * Generics use in order:
 *
 * - params in doInBackground()
 * - used during execution
 * - response in onPostExecute
 */
class SendContactsTask(
        private val context: Context): AsyncTask<Void, Void, String>() {

    //TODO: replace this Progress Dialog by something more user friendly
    var dialog: ProgressDialog? = null

    override fun doInBackground(vararg params: Void?): String? {
        val dao = ContactDAO(context)
        val contacts = dao.getContacts()
        dao.close()
        val json = Gson().toJson(contacts)
        return WebClient().post(json)
    }

    override fun onPreExecute() {
        dialog = ProgressDialog.show(context,context.getString(R.string.please_wait),context.getString(R.string.sending_contacts),true,true)
    }

    override fun onPostExecute(result: String?) {
        dialog?.dismiss()
        Toast.makeText(context,result,Toast.LENGTH_LONG).show()
    }
}
